ops/hookcmds/_relation.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import json
+from typing import (
+    Mapping,
+    cast,
+    overload,
+)
+
+from .._private import yaml
+from ._types import RelationModel, RelationModelDict
+from ._utils import run
+
+
+@overload
+def relation_get(
+    id: int | None = None,
+    *,
+    unit: str | None = None,
+    app: bool = False,
+) -> dict[str, str]: ...
+@overload
+def relation_get(
+    id: int | None = None,
+    *,
+    key: str,
+    unit: str | None = None,
+    app: bool = False,
+) -> str: ...
+def relation_get(
+    id: int | None = None,
+    *,
+    key: str | None = None,
```

Comment:
key name `-` is more special...

Reply:
I've changed it to explicitly disallow `-` as a key - if you want all the keys, then you have to pass `None` as the key (the default).

I think this is more Pythonic (only one obvious way...) and it means that we can continue to have a string value for `key` mean you get a string response, rather than a string value meaning either a string or a dictionary, which is nicer for the caller as well.